apps/vscode/command_client.py

```diff
@@ -1,126 +1,235 @@
     # the active VSCode instance will accept keypresses, we can be sure that
     # the active VSCode instance will be the one to write the port.
-    if is_mac:
-        actions.key("cmd-shift-alt-p")
-    else:
-        actions.key("ctrl-shift-alt-p")
+    actions.user.trigger_command_server_command_execution()
+
+    try:
+        decoded_contents = read_json_with_timeout(response_path)
+    finally:
+        unlink_if_exists(request_path)
+        unlink_if_exists(response_path)
+
+    if decoded_contents["uuid"] != uuid:
+        raise Exception("uuids did not match")
 
-    # Wait for the VSCode instance to update the port file.  This generally
-    # happens within the first millisecond, but we give it 3 seconds just in
-    # case.
-    start_time = time.monotonic()
-    new_contents = port_file_path.read_text()
+    if decoded_contents["error"] is not None:
+        raise Exception(decoded_contents["error"])
+
+    actions.sleep("25ms")
+
+    return decoded_contents["returnValue"]
+
+
+def unlink_if_exists(path):
+    try:
+        path.unlink()
+    except FileNotFoundError:
+        pass
+
+
+def read_json_with_timeout(path: str) -> Any:
+    """Repeatedly tries to read a json object from the given path, waiting until there is a trailing new line indicating that the write is complete
+
+    Args:
+        path (str): The path to write to
+
+    Raises:
+        Exception: If we timeout waiting for a response
+
+    Returns:
+        Any: The json-decoded contents of the file
+    """
+    start_time = time.perf_counter()
     sleep_time = 0.0005
     while True:
-        if new_contents != original_contents:
-            try:
-                decoded_contents = json.loads(new_contents)
-                # If we're successful, we break out of the loop
+        try:
+            raw_text = path.read_text()
+
+            if raw_text.endswith("\n"):
                 break
-            except ValueError:
-                # If we're not successful, we keep waiting; we assume it was a
-                # partial write from VSCode
-                pass
-        time.sleep(sleep_time)
-        sleep_time *= 2
-        if time.monotonic() - start_time > 3.0:
-            raise Exception("Timed out waiting for VSCode to update port file")
-        new_contents = port_file_path.read_text()
-
-    port = decoded_contents["port"]
-    nonce = decoded_contents["nonce"]
-
-    response = requests.post(
-        f"http://localhost:{port}/execute-command",
-        json={
-            "commandId": command,
-            "args": args,
-            "waitForFinish": wait_for_finish,
-            "expectResponse": expect_response,
-            "nonce": nonce,
-        },
-        timeout=(0.05, 3.05),
-    )
-    response.raise_for_status()
+        except FileNotFoundError:
+            # If not found, keep waiting
+            pass
 
-    actions.sleep("25ms")
+        actions.sleep(sleep_time)
+        sleep_time *= 2
```

Comment:
You might want to clamp the value of `sleep_time` so you don't overshoot the timeout.
Perhaps, before the loop, set `timeout_time = now + 3.0`.
Then clamp sleep time to something like this after doubling it:
```
sleep_time *= 2
time_left = timeout_time - time.perf_counter()
sleep_time = max(0.0005, min(sleep_time, time_left))
```

Otherwise, you can only timeout after one of these total times:

```
acc = 0
sleep_time = 0.0005
for i in range(15):
    acc += sleep_time
    sleep_time *= 2
    print(f"{i} sleep={sleep_time:.3f}s total={acc:.3f}s")
```

```
0 sleep=0.001s total=0.001s
1 sleep=0.002s total=0.002s
2 sleep=0.004s total=0.004s
3 sleep=0.008s total=0.007s
4 sleep=0.016s total=0.015s
5 sleep=0.032s total=0.032s
6 sleep=0.064s total=0.064s
7 sleep=0.128s total=0.128s
8 sleep=0.256s total=0.256s
9 sleep=0.512s total=0.512s
10 sleep=1.024s total=1.024s
11 sleep=2.048s total=2.048s
12 sleep=4.096s total=4.096s
13 sleep=8.192s total=8.192s
14 sleep=16.384s total=16.384s
```